### PR TITLE
Fix rollback failure when deploying Deployment through yaml

### LINKED ISSUE #7804: [BUG] Problem with deployment Rollback through the Rancher GUI - Deployment.apps is invalid: spec.template.metadata.labels: Invalid value: map[string]string{}: selector does not match template labels
**Rancher Server Setup**
- Rancher version:
 v2.6.8

- Installation option (Docker install/Helm Chart):
Docker install

**Information about the Cluster**
- Kubernetes version:
1.21

- Cluster Type (Local/Downstream):
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):
Hosted, Imported from EKS

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  Admin

**Describe the bug**
When I try to rollback I get the following message:
`Deployment.apps “api-notifies-php” is invalid: spec.template.metadata.labels: Invalid value: map[string]string{}: selector does not match template labels`

**To Reproduce**
    1. From the Global view, open the project running the workload you want to rollback.
    2. Find the workload that you want to rollback and select Vertical ⋮ (... ) > Rollback.
    3. Choose the revision that you want to roll back to. Click Rollback.

**Result**
Error:
`Deployment.apps “api-notifies-php” is invalid: spec.template.metadata.labels: Invalid value: map[string]string{}: selector does not match template labels`

**Expected Result**
The expected rollback is successful.

**Screenshots**
https://ibb.co/kcn3nfC

**Additional context**
I’ve reviewed the Deployment manifest, adjusted the indentation, made sure the matchLabels selector is the same as the Pod labels in the templates section.
I tried several things, but I couldn’t solve this problem, in order to perform the Deployment Rollback.
Through kubectl the rollback is working correctly.

Here is my deployment manifest:

```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: teste-devops
  labels:
    app: nginx
spec:
  template:
    metadata:
      labels:
        projeto: testando
    spec:
      containers:
      - name: nginx
        image: nginx:1.20.0
        ports:
        - containerPort: 80
  replicas: 2
  selector:
    matchLabels:
      projeto: testando
```

Docker version(on EC2 instance where Rancher is hosted):

```

Client: Docker Engine - Community
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:02:57 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:03 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

Details about the Kubernetes nodes inside Rancher Container:

```
sh-4.4# kubectl get nodes
NAME         STATUS   ROLES                       AGE    VERSION
local-node   Ready    control-plane,etcd,master   132m   v1.24.1+k3s1
sh-4.4# 
```

### PR DESCRIPTION
**Pull request by huacheng-git**
### Summary
Fixes [#37928](https://github.com/rancher/dashboard/issues/7803)
Fixes #7804

### CODE REVIEW

File: shell/models/apps.deployment.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import { WORKLOAD_TYPES } from '@shell/config/types';
 import Workload from './workload';
+import pickBy from 'lodash/pickBy';
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply shell/models/apps.deployment.js:3]: you might also be able to do something like `const { 'pod-template-hash', ...otherLabels } = revision.spec.template.metadata.labels`. 

```
const a = { a:1, b:2, c:3 }; 
undefined
const {  a, ...others } = a
undefined
others
{b: 2, c: 3}
```

File: shell/models/apps.deployment.js [hunk 2]
```diff
@@ -19,7 +20,8 @@ export default class Deployment extends Workload {
         value: {
           metadata: {
             creationTimestamp: null,
-            labels:            { 'workload.user.cattle.io/workloadselector': revision.spec.template.metadata.labels['workload.user.cattle.io/workloadselector'] }
+            labels:            pickBy(revision.spec.template.metadata.labels, (value, key) => key !== 'pod-template-hash'),
+            annotations:       { 'cattle.io/timestamp': revision.spec.template.metadata.annotations['cattle.io/timestamp'] },
```
Comment (richard-cox) [inline shell/models/apps.deployment.js:24]: out of interest do we need this, what happens if the timestamp annotation  isn't there in the revision

**Review by richard-cox — CHANGES_REQUESTED** [review]

File: shell/models/apps.deployment.js (hunk 2 above)
Comment (huacheng-git) [reply shell/models/apps.deployment.js]: If there is no timestamp, the normal old Pod will be removed during rollback.
To Reproduce
1. Create Deployment app, replica 1, image busybox, stdin yes, tty.
2. Click Redeploy on the UI, cattle.io/timestamp will be generated in spec.template.metadata.annotations.
3. Edit Deployment app，image busybox，no stdin yes，no tty。
4. At this time, the new Pod of the Deployed app cannot be created successfully, and the old Pod is running.
5. Rollback Deployment app，the old running Pod is also deleted.

There are still some problems in this way. I have made some adjustments.For spec.template.metadata.annotations, need to copy other items, refer to https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/rollback.go#L99-L166
The annotationsToSkip needs to be ignored, and the other annos need to be copied

**Review by richard-cox — APPROVED** [review]
LGTM

I took a while to work out how `kubectl rollout undo deployment/[deployment name] --to-revision=[revision number] -v=8` behaves and we should be doing as this PR does (copying over spec labels/annotations in the patch request)

**Pull request merged**